### DISCUSSION TITLE
RuntimeError: Trying to backward through the graph a second time (Section 02)

### DISCUSSION THREAD

**Discussion by hhtt67678** [mrdbourke/pytorch-deep-learning#334] [Q&A] [answered]
I'm working along with chapter 02 and am writing a model with 3 nn.Linear layers (as done in the video). I'm following along pretty closely to the video, but I tried to write the model and loop on my own to test whether I could do it. When running the training loop I am getting this error at the `loss.backward()` step in the training loop:

> RuntimeError: Trying to backward through the graph a second time (or directly access saved tensors after they have already been freed). Saved intermediate values of the graph are freed when you call .backward() or autograd.grad(). Specify retain_graph=True if you need to backward through the graph a second time or if you need to access saved tensors after calling backward.

I searched on StackOverflow, Google, and this forum, and though this error comes up frequently, none of the answers appeared to apply to my specific, very simple model. I also tried adding `retain_graph=True` as stated in the error text but that also didn't work.

Model Code:
```
class CircleModelV1(nn.Module):
  def __init__(self):
    super().__init__()
    self.layer_1 = nn.Linear(in_features=2,
                             out_features=10)
    self.layer_2 = nn.Linear(in_features = 10,
                             out_features=10)
    self.layer_3 = nn.Linear(in_features=10,
                             out_features=1)
  
  def forward(self, x):
    return self.layer_3(self.layer_2(self.layer_1(x))) # speedier this way for some reason
```

Training loop code:

```
epochs= 1000


X_train, X_test = X_train.to(device), X_test.to(device)
y_train, y_test = y_train.to(device), y_test.to(device)

for epoch in range(epochs):

  model_2.train()

  y_train_logits = model_2(X_train).squeeze()
  y_pred = torch.round(torch.sigmoid(y_train_logits))


  loss = loss_fn(y_logits, y_train)
  acc = accuracy_fn(y_true=y_train,
                    y_pred=y_pred)


  optimizer.zero_grad()


  loss.backward() # <- This step throws an error


  optimizer.step()


  model_2.eval()

  with torch.inference_mode():

    y_test_logits = model_2(X_test).squeeze()
    y_test_pred = torch.round(torch.sigmoid(y_test_logits))


    test_loss_logits = loss_fn(y_test_logits, y_test)
    test_acc = accuracy_fn(y_true=y_test, y_pred=y_test_pred)

    if epoch % 20 == 0:
      print(f'Epoch: {epoch}, Loss:{loss:.4f}, Test Loss:{test_loss:.4f}, Test Accuracy:{test_acc:.2f}')
```

I would greatly appreciative of any suggestions of where to look!

Edit: Still getting same error when using `nn.Sequential` to write the model. So I know the error has something to do with putting hidden layers into the model, but beyond that not sure. Continuing to search for answers...

**Comment by mrdbourke** [comment, marked as answer]
Hi @hhtt67678,

Epic work trying to create neural networks of your own!

Looks like your error might be with this line of code: `loss = loss_fn(y_logits, y_train)`

You're trying to calculate loss on `y_logits` but your `model_2` outputs `y_train_logits`.

Before:

```python
y_train_logits = model_2(X_train).squeeze()
y_pred = torch.round(torch.sigmoid(y_train_logits))

loss = loss_fn(y_logits, y_train)
acc = accuracy_fn(y_true=y_train,
                  y_pred=y_pred)
```

After (`y_train_logits` passed to `loss_fn` instead of `y_logits`): 

```python
y_train_logits = model_2(X_train).squeeze()
y_pred = torch.round(torch.sigmoid(y_train_logits))


loss = loss_fn(y_train_logits, y_train) # <- changed this line
acc = accuracy_fn(y_true=y_train,
                  y_pred=y_pred)
```

Your full code (updated):

```python
epochs= 1000


X_train, X_test = X_train.to(device), X_test.to(device)
y_train, y_test = y_train.to(device), y_test.to(device)

for epoch in range(epochs):

  model_2.train()

  y_train_logits = model_2(X_train).squeeze()
  y_pred = torch.round(torch.sigmoid(y_train_logits))


  loss = loss_fn(y_train_logits, y_train)
  acc = accuracy_fn(y_true=y_train,
                    y_pred=y_pred)


  optimizer.zero_grad()


  loss.backward() # <- This step throws an error


  optimizer.step()


  model_2.eval()

  with torch.inference_mode():

    y_test_logits = model_2(X_test).squeeze()
    y_test_pred = torch.round(torch.sigmoid(y_test_logits))


    test_loss_logits = loss_fn(y_test_logits, y_test)
    test_acc = accuracy_fn(y_true=y_test, y_pred=y_test_pred)

    if epoch % 20 == 0:
      print(f'Epoch: {epoch}, Loss:{loss:.4f}, Test Loss:{test_loss:.4f}, Test Accuracy:{test_acc:.2f}')
```

Let me know how this goes!

**Reply by hhtt67678** [reply]
That did the trick, thank you! I'm relatively new to this and still figuring out where to look for these kinds of errors. BTW I am enjoying learning from your videos, thank you for the time you've put into them!
